### PR TITLE
token type ids can be set by optional argument up to python wrapper

### LINKED ISSUE #1383: token_type_ids for BERT models
Hi, we have run into some problems when working with Ctranslate for BERT models, as for a number of tasks such as question answering and sentence pair comparison, they have been trained with token_type_ids as part of their inputs. 

Request:
We would like to inquire if there is a plan to address this limitation in the near future. If there is no such plan currently, we would like to express our willingness to contribute to the project by writing a patch that introduces support for token_type_ids.

Benefits:
* Improved BERT model performance by handling token_type_ids.
* Increased compatibility with tasks that require sentence pairs or multiple input sequences.
* Expanded usage of the library for a broader range of natural language processing tasks.

Would you consider accepting a patch for this issue if we were to implement it? We would be more than happy to collaborate with the community and follow any necessary guidelines to ensure that the patch aligns with the project's standards.

### PR DESCRIPTION
**Pull request by hachall**
# Description

This PR adds an optional token_type_ids argument to Encoder.translate_batch, to facilitate BERT models with multiple input features (sentence-sentence entailment, context and question answering etc.). The token_type_ids vector is optional at the pybind level, and overloaded into the future_batch_async functions, with all previous endpoints preserved for backwards compatibility.

Closes #1383.

```
...
output = ct2_model.forward_batch(input_ids, token_type_ids=token_type_ids)
pooler_output = output.pooler_output
pooler_output = np.array(pooler_output)
pooler_output = torch.as_tensor(pooler_output)
ct_logits = classifier(pooler_output)
```

# Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Docs update
- [ ] Refactor

# Testing
The code passes all C++ and python tests. For validation, the modified encoder was tested with both a bert-base-uncased and fine tuned bert-tiny for inference on the train split of the MRPC dataset (sentence pairs), and all output logits verified equal (<10e-12) to the output of a hugging face loaded model at full precision, when passing token_type_ids to both. When quantized, the logits are no longer equal, but the token_type_ids demonstrably improve the classifier accuracy. 

# Checklist:

- [X] My code follows the style guidelines of this project
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by hachall** [comment]
#1383

File: include/ctranslate2/encoder.h [hunk 1]
```diff
@@ -13,11 +13,20 @@ namespace ctranslate2 {
...
     forward_batch_async(std::vector<std::vector<size_t>> ids);
 
+    std::future<EncoderForwardOutput>
+    forward_batch_async(std::vector<std::vector<size_t>> ids, std::vector<std::vector<size_t>> token_type_ids);
+
     std::future<EncoderForwardOutput>
     forward_batch_async(const StorageView& ids, const StorageView& lengths);
+    
+    std::future<EncoderForwardOutput>
+    forward_batch_async(const StorageView& ids, const StorageView& lengths, std::vector<std::vector<size_t>> token_type_ids);
```
Comment (guillaumekln) [inline include/ctranslate2/encoder.h:29]: I think we can avoid creating new method overloads by adding the new argument with a default value to the existing methods. Something like:

```cpp
std::future<EncoderForwardOutput>
forward_batch_async(std::vector<std::vector<std::string>> tokens,
                    std::vector<std::vector<size_t>> token_type_ids = {});
```

What do you think?

File: include/ctranslate2/models/language_model.h [hunk 2]
```diff
@@ -124,12 +124,15 @@ namespace ctranslate2 {
       }
 
       EncoderForwardOutput forward(const std::vector<std::vector<std::string>>& tokens);
+      EncoderForwardOutput forward(const std::vector<std::vector<std::string>>& tokens, const std::vector<std::vector<size_t>>& token_type_ids);
       EncoderForwardOutput forward(const std::vector<std::vector<size_t>>& ids);
+      EncoderForwardOutput forward(const std::vector<std::vector<size_t>>& ids, const std::vector<std::vector<size_t>>& token_type_ids);
       EncoderForwardOutput forward(const StorageView& ids, const StorageView& lengths);
+      EncoderForwardOutput forward(const StorageView& ids, const StorageView& lengths, const std::vector<std::vector<size_t>>& token_type_ids);
```
Comment (guillaumekln) [inline include/ctranslate2/models/language_model.h:131]: Same comment regarding adding the argument with a default value, instead of overloading the methods.

File: python/cpp/encoder.cc [hunk 3]
```diff
@@ -81,8 +86,8 @@ namespace ctranslate2 {
                    device: Device to use (possible values are: cpu, cuda, auto).
                    device_index: Device IDs where to place this encoder on.
                    compute_type: Model computation type or a dictionary mapping a device name
-                     to the computation type (possible values are: default, auto, int8, int8_float32,
-                     int8_float16, int8_bfloat16, int16, float16, bfloat16, float32).
+                     to the computation type (possible values are: default, auto, int8, int8_float16,
+                     int8_bfloat16, int16, float16, bfloat16, float32).
```
Comment (guillaumekln) [inline python/cpp/encoder.cc:90]: Revert this change (probably a bad merge).

File: python/cpp/encoder.cc [hunk 4]
```diff
@@ -97,8 +102,6 @@ namespace ctranslate2 {
                                "Device this encoder is running on.")
         .def_property_readonly("device_index", &EncoderWrapper::device_index,
                                "List of device IDs where this encoder is running on.")
-        .def_property_readonly("compute_type", &EncoderWrapper::compute_type,
-                               "Computation type used by the model.")
```
Comment (guillaumekln) [inline python/cpp/encoder.cc:101]: Revert this change (probably a bad merge).

File: python/cpp/encoder.cc [hunk 5]
```diff
@@ -119,6 +123,8 @@ namespace ctranslate2 {
                      ``[batch_size, max_length]`` (e.g. created from a Numpy array or PyTorch tensor).
                    lengths: The length of each sequence as a int32 array with shape
                      ``[batch_size]``. Required when :obj:`inputs` is a dense array.
+                   token_type_ids: A batch of token type IDs of same shape as :obj:`inputs`.
+                      ``[batch_size, max_length]``.
```
Comment (guillaumekln) [inline python/cpp/encoder.cc:127]: Minor indentation fix for consistency.

```suggestion
                     ``[batch_size, max_length]``. 
```

File: src/models/language_model.cc [hunk 6]
```diff
@@ -360,9 +379,13 @@ namespace ctranslate2 {
...
       if (_encoder->num_input_features() > 1) {
-        StorageView token_type_ids(ids.shape(), ids.dtype(), device);
-        token_type_ids.zero();
-        inputs.emplace_back(std::move(token_type_ids));
+        if (token_type_ids.empty()) {
+          StorageView placeholder_type_ids(ids.shape(), ids.dtype(), device);
+          placeholder_type_ids.zero();
+          inputs.emplace_back(std::move(placeholder_type_ids));
+        } else {
+          inputs.emplace_back(std::move(token_type_ids));
```
Comment (guillaumekln) [inline src/models/language_model.cc:387]: `token_type_ids` is a const reference so the `std::move` has no effect.

```suggestion
          inputs.emplace_back(token_type_ids);
```

File: src/models/language_model.cc [hunk 7]
```diff
@@ -298,30 +298,49 @@ namespace ctranslate2 {
...
+    }
+    
+    EncoderForwardOutput
+    SequenceEncoderReplica::forward(const StorageView& ids, const StorageView& lengths, const std::vector<std::vector<size_t>>& token_type_ids) {
       PROFILE("SequenceEncoderReplica::forward");
       const auto& model = *this->model();
       const auto device = model.device();
       const auto scoped_device_setter = model.get_scoped_device_setter();
 
+      StorageView input_token_type_ids = layers::make_sequence_inputs(token_type_ids, Device::CPU, 1, nullptr);
```
Comment (guillaumekln) [inline src/models/language_model.cc:337]: Directly create the storage on the target device.

```suggestion
      StorageView input_token_type_ids = layers::make_sequence_inputs(token_type_ids, device);
```

File: src/models/language_model.cc [hunk 8]
```diff
@@ -298,30 +298,49 @@ namespace ctranslate2 {
...
       EncoderForwardOutput output;
 
       if (ids.device() != device)
-        output = forward_impl(ids.to(device), lengths.to(device));
+        output = forward_impl(ids.to(device), lengths.to(device), input_token_type_ids.to(device));
```
Comment (guillaumekln) [inline src/models/language_model.cc:341]: A `.to(device)` is no longer needed if the input is created on the correct device.

```suggestion
        output = forward_impl(ids.to(device), lengths.to(device), input_token_type_ids);
```

File: python/cpp/encoder.cc [hunk 9]
```diff
@@ -13,20 +13,25 @@ namespace ctranslate2 {
...
+                    const std::optional<StorageView>& lengths,
+                    const std::optional<BatchIds>& token_type_ids
+                    ) {
         std::future<EncoderForwardOutput> future;
 
         switch (inputs.index()) {
         case 0:
-          future = _pool->forward_batch_async(std::get<BatchTokens>(inputs));
+          future = (token_type_ids) ? _pool->forward_batch_async(std::get<BatchTokens>(inputs), token_type_ids.value()) 
+                                    : _pool->forward_batch_async(std::get<BatchTokens>(inputs));
```
Comment (guillaumekln) [inline python/cpp/encoder.cc:24]: Is it possible to avoid duplicating the method call? Something like this:

```cpp
future = _pool->forward_batch_async(std::get<BatchTokens>(inputs), token_type_ids.value_or(std::vector<std::vector<size_t>>()));
```
